File: aoc-2022-kotlin/src/main/kotlin/days/day01/day01.kt
package days.day01

import lib.Reader
import lib.checkAnswer
import lib.time
import kotlin.math.max

fun main() {
    val input = Reader("day01.txt").strings()
    val exampleInput = Reader("day01-example.txt").strings()

    time(message = "Part 1") {
        part1(input)
    }.checkAnswer(67450)

    time(message = "Part 2") {
        part2(input)
    }.checkAnswer(199357)
}

fun part1(lines: List<String>): Int {
    var max = 0
    var runningTotal = 0
    for (line in lines) {
        if (line.isNullOrEmpty()) {
            if (runningTotal > max) max = runningTotal
            runningTotal = 0
        } else {
            runningTotal += line.toInt()
        }
    }
    return max(runningTotal, max)
}

fun part2(lines: List<String>): Int {
    val elves = mutableListOf<Int>()
    var runningTotal = 0
    for (line in lines) {
        if (line.isNullOrEmpty()) {
            elves.add(runningTotal)
            runningTotal = 0
        } else {
            runningTotal += line.toInt()
        }
    }
    elves.add(runningTotal)
    elves.sortDescending()
    return elves.take(3).sum()
}

File: aoc-2019-go/day01/day01.go
package main

import (
	"fmt"
	"reader"
)

func main() {
	input := reader.Ints("./input.txt")
	fmt.Println(part1(input))
	fmt.Println(part2(input))
}

func part1(masses []int) int {
	sum := 0
	for _, mass := range masses {
		sum = sum + getFuel(mass)
	}
	return sum
}

func part2(masses []int) int {
	sum := 0
	for _, mass := range masses {
		sum = sum + getFuel2(mass)
	}
	return sum
}

func getFuel(mass int) int {
	return mass/3 - 2
}

func getFuel2(mass int) int {
	fuel := getFuel(mass)
	if fuel > 0 {
		return fuel + getFuel2(fuel)
	} else {
		return 0
	}
}
